util/mobile_base.hpp

```diff
@@ -0,0 +1,325 @@
+      base_trajectories_.pop();
+    }
+
+    // go into compliance mode if no trajectory
+    if (base_trajectories_.size() == 0) {
+      auto size = group_manager_->size();
+
+      VectorXd compliantState;
+      compliantState.resize(size);
+      double nan = std::numeric_limits<double>::quiet_NaN();
+      for (auto i=0; i<size; ++i) {
+        compliantState(i) = nan;
+      }
+
+      cmd.setPosition(compliantState);
+      cmd.setVelocity(compliantState);
+      compliantState.setZero();
+      cmd.setEffort(compliantState);
+
+      return ret;
+    } else {
+      auto traj = base_trajectories_.front();
+
+      VectorXd pos, vel, accel;
+
+      // Update command from trajectory
+      traj->getState(group_manager_->lastTime(), &pos, &vel, &accel);
+
+      // set velocity to steer along cartesian trajectory
+
+      // Convert from x/y/theta to wheel velocities
+      auto wheel_vels = SE2ToWheelVel({pos(0), pos(1), pos(2)}, {vel(0), vel(1), vel(2)});
+
+      // Integrate position using wheel velocities.
+      last_wheel_pos_ += wheel_vels * group_manager_->dT();
+      cmd.setPosition(last_wheel_pos_);
+
+      // Use velocity from trajectory, converted from x/y/theta into wheel velocities above.
+      cmd.setVelocity(wheel_vels);
+
+      return ret;
+    }
+  }
+
+  bool send() {
+    return group_manager_->send();
+  }
+
+  //virtual SE2Point wheelsToSE2(Eigen::VectorXd wheel_state) const = 0;
+  virtual Eigen::VectorXd SE2ToWheelVel(Pose pos, Vel vel) const = 0;
+
+  // Get the global odometry state
+  Pose getOdometry() const { return global_pose_ + relative_odom_offset_; };
+
+  // Reset the current odometry state
+  void setOdometry(Pose p) {
+    last_wheel_pos_ = group_manager_->lastFeedback().getPosition();
+    relative_odom_offset_ = p - getOdometry();
+  }
+
+  // Get the maximum velocity in each direction
+  // NOTE: consider whether this is guaranteed maximum velocity
+  // or maximum velocity if no other directions are used!  Also
+  // is this true for any config of legged robots, or is this some
+  // "safe" max vel?
+  // TODO: potential make this depend on state?
+  virtual Vel getMaxVelocity() const = 0;
+
+  // Set the goal.  Goal will be actively commanded until cleared,
+  // maintaining the end state of the goal after finishing a smooth
+  // trajectory to the goal.
+  // Note -- this turns the `Goal` into a `Trajectory` based on the
+  // `getMaxVelocity` function implementation
+  // TODO: "best effort" -- or return "can't do this"?
+  bool setGoal(const CartesianGoal& g) {
+    auto baseTrajectories = buildTrajectory(g);
+    if (baseTrajectories) {
+      base_trajectories_ = *baseTrajectories;
+      return true;
+    }
+    return false;
+  }
+
+  // When cleared, the base should be passive / compliant if possible.
+  void clearGoal() {
+    // Taken from https://stackoverflow.com/a/709161
+    // clear the trajectory queue
+    std::queue<shared_ptr<Trajectory>>().swap(base_trajectories_);
```

Comment:
this might work? `std::erase(std::begin(base_trajectories_), std::end(base_trajectories_));`